### ISSUE TITLE
Error after 99% of converting video

### ISSUE THREAD

**Issue by Isti115** [Franiac/TwitchLeecher#272]
I tried to download this video: https://www.twitch.tv/videos/237704174
Using Twitch Leecher 1.5.2
On Windows 10 (x64)

And after successfully downloading and merging it the conversion phase failed after reaching around 99.1%... (In this case if I may, I would suggest leaving the temporary folder in place at least until the program is open, so I could've at least salvaged the *.ts file and convert it myself!)

(This was my second trial, I wasn't watching the first run, just found the error after it, so I tried again to see exactly what happens.)

Here is the log:

> Download task has been started!
> 
> TWITCH LEECHER INFO
> --------------------------------------------------------------------------------------------
> Version: 1.5.2
> 
> VOD INFO
> --------------------------------------------------------------------------------------------
> VOD ID: 237704174
> Selected Quality: 1920x1080@25fps (Source)
> Download Url: https://www.twitch.tv/videos/237704174
> Crop Start: No
> Crop End: No
> 
> OUTPUT INFO
> --------------------------------------------------------------------------------------------
> Output File: E:\Twitch\20180311_237704174_Command & Conquer_ Red Alert 3.mp4
> FFMPEG Path: C:\Program Files\Twitch Leecher\ffmpeg_x64.exe
> Temporary Download Folder: C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688
> 
> ACCESS INFO
> --------------------------------------------------------------------------------------------
> Token: %7B%22authorization%22%3A%7B%22forbidden%22%3Afalse%2C%22reason%22%3A%22%22%7D%2C%22chansub%22%3A%7B%22restricted_bitrates%22%3A%5B%5D%7D%2C%22device_id%22%3Anull%2C%22expires%22%3A1530735302%2C%22https_required%22%3Afalse%2C%22privileged%22%3Afalse%2C%22user_id%22%3Anull%2C%22version%22%3A2%2C%22vod_id%22%3A237704174%7D
> Signature: 450c92f38c482d289bfde15db5c68ebee6ed93a2
> Sub-Only: No
> Privileged: No
> 
> Creating temporary download directory 'C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688'... done!
> 
> Retrieving m3u8 playlist urls for all VOD qualities... done!
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/chunked/highlight-237704174-muted-WT9OGDRR05.m3u8
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/720p30/highlight-237704174-muted-WT9OGDRR05.m3u8
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/480p30/highlight-237704174-muted-WT9OGDRR05.m3u8
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/audio_only/highlight-237704174-muted-WT9OGDRR05.m3u8
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/360p30/highlight-237704174-muted-WT9OGDRR05.m3u8
> https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/160p30/highlight-237704174-muted-WT9OGDRR05.m3u8
> 
> Playlist url for selected quality 1920x1080@25fps (Source) is https://fastly.vod.hls.ttvnw.net/f8325ac45e9aec6986d9_gbhmichael_27894237024_812059978/chunked/highlight-237704174-muted-WT9OGDRR05.m3u8
> 
> Retrieving playlist... done!
> Parsing playlist... done!
> Number of video chunks: 2632
> 
> Starting parallel video chunk download
> Number of video chunks to download: 2632
> Maximum connection count: 10
> 
> Parallel video chunk download is running...
> 
> Download of all video chunks complete!
> 
> Merging all VOD parts into 'C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688\20180311_237704174_Command & Conquer_ Red Alert 3.ts'...
> 
> Executing 'C:\Program Files\Twitch Leecher\ffmpeg_x64.exe' on 'C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688\20180311_237704174_Command & Conquer_ Red Alert 3.ts'...
> Command line arguments: -y -i "C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688\20180311_237704174_Command & Conquer_ Red Alert 3.ts" -analyzeduration 2147483647 -probesize 2147483647 -c:v copy -c:a copy -bsf:a aac_adtstoasc "E:\Twitch\20180311_237704174_Command & Conquer_ Red Alert 3.mp4"
> 
> 
> Starting temporary download folder cleanup!
> Deleting directory 'C:\Users\isti\AppData\Local\Temp\TL_6818d9a4-702d-4408-bb82-913b95206688'... done!
> 
> Download task ended with an error!
> 
> System.AggregateException: One or more errors occurred. ---> System.ApplicationException: An error occured while converting the video!
>    at TwitchLeecher.Services.Services.ProcessingService.ConvertVideo(Action`1 log, Action`1 setStatus, Action`1 setProgress, Action`1 setIsIndeterminate, String sourceFile, String outputFile, CropInfo cropInfo)
>    at TwitchLeecher.Services.Services.TwitchService.<>c__DisplayClass57_0.<StartQueuedDownloadIfExists>b__2()
>    at System.Threading.Tasks.Task.Execute()
>    --- End of inner exception stack trace ---
> ---> (Inner Exception #0) System.ApplicationException: An error occured while converting the video!
>    at TwitchLeecher.Services.Services.ProcessingService.ConvertVideo(Action`1 log, Action`1 setStatus, Action`1 setProgress, Action`1 setIsIndeterminate, String sourceFile, String outputFile, CropInfo cropInfo)
>    at TwitchLeecher.Services.Services.TwitchService.<>c__DisplayClass57_0.<StartQueuedDownloadIfExists>b__2()
>    at System.Threading.Tasks.Task.Execute()<---
>

**Comment by Franiac** [comment]
If ffmpeg fails to process the video I really cannot do anything about it. It is an indicator that the source material has some kind of faulted frame in it that ffmpeg does not like (streamer dropped frames for example). If that happend my hands are tied.

Keeping the file is something that I do not want to do because of the following reasons:

1. The user  should not care about anything the a program does in the background. Either it works or not. 99.9% of the users do not know what a *.ts file even is.
2. If ffmpeg fails to process the ts file, how big are the chances any other video software can (since many of them rely on ffmpeg)?
3. Keeping the file until the user removes the download in TL could be done but it has one major flaw: If you close TL and there are 5 faulted downloads with let's say 200gb of temp files laying around, those must be cleaned up when shutting down TL. On slow machines shutting down would be a 1 min or longer task. Such a behaviour is unacceptable.

**Comment by Isti115** [comment]
Wow, such a quick and well written answer!
Thank you very much for your reply!
It's nice to see someone provide quality support for their hobby project! 👍

I will have to admit that those are indeed some valid reasons, but I would still either consider adding an advanced option in the settings to keep the `*.ts` file, or maybe prompt the user in case of an encoding error whether they would like to keep it,

So you are sure that this error is caused by ffmpeg not being able to process the file? I saw no clear indication of this in the log, maybe in this case you could print the actual message provided by ffmpeg as well. (Just a suggestion. 😉)

**Comment by Isti115** [comment]
tl;dr:
### Check the available storage space and the file system of your destination!!!

---

Hmm, I have tried to download the video again in hopes of copying the `*.ts` file from the temporary folder to a "safe place" while it was trying to encode, and succeeded doing so. After that I started to reencode the whole file into mp4 using ffmpeg that was (as expected) a horribly slow process, it was chewing through it at around 2x to 3x speed.

Meanwhile having left Twitch Leecher open I have somehow noticed that the status text was green this time and read "Done". In my surprise I checked the location I would've liked to download it and was even more surprised when I found it to be completely empty. At this point I clicked the "Open Download Folder" button and figured out that I must've forgotten to override the download path as I did yesterday, so it put it into the default Downloads folder. I tried to copy it onto the (64 GB) SD card<sup>1</sup> where I had originally intended for it to go and BAM, Windows told me that the file was too big for the file system... The problem was that the SD card was FAT32 the whole time... So the moral of the story is to check the destination file system as well as the remaining space there.

Finally I have managed to split the video onto the card using the second command as a template from here: https://superuser.com/a/140980/158498

1: I will be travelling a bit next week to somewhere with not Wi-Fi, that's why I wanted to download the VOD in the first place, but I didn't want to bring an external HDD so I figured that a huge SD card will do.

**Comment by Franiac** [comment]
If TL fails at the processing stage it is 100% a ffmpeg error because that's all that stage does. The ffmpeg error message is hidden intentionally. In the worst case ffmpeg outputs 5 lines of log text for every processed frame which can add up to millions of lines of text. And even if you had the error message... what would you do? Have you ever seend a common ffmpeg error message? HAVE FUN if you are not a human encoder yourself 😄 

Still I am not convinced that keeping the ts file (even optionally) is a good thing to implement because it is an option 0.01% of all users need or even understand the reason behind it. TL is aimed at gamers who want to download videos from Twitch, not software architects who know every command line parameter of ffmpeg (without saying gamers are dumb people..... I am a gamer myself).

Conclusion: I will not take any action out of this because it is a very rare problem and I rate the possible solution as not useful for the majority of my users. I hope you can live with that. If you encounter the issue more often than not, you surely can come back and re-open.
